gnmi_server/gnoi_system_test.go

```diff
@@ -0,0 +1,261 @@
+			t.Errorf("nsfManagerResponse failed to send response: %v", err)
+			return
+		}
+	case <-done:
+		return
+	case <-tc:
+		t.Error("nsfManagerResponse timed out waiting for request")
+		return
+	}
+}
+
+func TestGnoiSystem(t *testing.T) {
+	s := createServer(t)
+	go runServer(t, s)
+
+	tlsConfig := &tls.Config{InsecureSkipVerify: true}
+	opts := []grpc.DialOption{grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig))}
+	targetAddr := fmt.Sprintf("127.0.0.1:%d", s.config.Port)
+	conn, err := grpc.Dial(targetAddr, opts...)
+	if err != nil {
+		t.Fatalf("Dialing to %s failed: %v", targetAddr, err)
+	}
+	defer conn.Close()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	sc := syspb.NewSystemClient(conn)
+	rclient, err := getRedisDBClient(stateDB)
+	if err != nil {
+		t.Fatalf("Cannot connect to the redis server: %v.", err.Error())
+	}
+	defer rclient.Close()
+
+	t.Run("RebootFailsIfRebootMethodIsNotSupported", func(t *testing.T) {
+		req := &syspb.RebootRequest{
+			Delay:   0,
+			Message: "Cold reboot due to ...",
+		}
+
+		for _, method := range []syspb.RebootMethod{syspb.RebootMethod_UNKNOWN, syspb.RebootMethod_HALT, syspb.RebootMethod_POWERUP} {
+			req.Method = method
+			_, err := sc.Reboot(ctx, req)
+			testErr(err, codes.InvalidArgument, "reboot method is not supported.", t)
+		}
+	})
+	t.Run("RebootFailsIfItIsDelayed", func(t *testing.T) {
+		req := &syspb.RebootRequest{
+			Method:  syspb.RebootMethod_COLD,
+			Delay:   1,
+			Message: "Cold reboot due to ...",
+		}
+
+		_, err := sc.Reboot(ctx, req)
+		testErr(err, codes.InvalidArgument, "reboot is not immediate.", t)
+	})
+	t.Run("RebootFailsIfMessageIsNotSet", func(t *testing.T) {
+		req := &syspb.RebootRequest{
+			Method: syspb.RebootMethod_COLD,
+			Delay:  0,
+		}
+
+		_, err := sc.Reboot(ctx, req)
+		testErr(err, codes.InvalidArgument, "message is empty.", t)
+	})
+	t.Run("RebootFailsWithTimeout", func(t *testing.T) {
+		req := &syspb.RebootRequest{
+			Delay:   0,
+			Message: "Reboot due to ...",
+		}
+
+		for _, method := range []syspb.RebootMethod{syspb.RebootMethod_COLD, syspb.RebootMethod_POWERDOWN, syspb.RebootMethod_WARM} {
+			req.Method = method
+			_, err := sc.Reboot(ctx, req)
+			testErr(err, codes.Internal, "Response Notification timeout from NSF Manager!", t)
+		}
+	})
+	t.Run("RebootFailsForInvalidOptics", func(t *testing.T) {
+		req := &syspb.RebootRequest{
+			Method:  syspb.RebootMethod_COLD,
+			Delay:   0,
+			Message: "Reset optics due to ...",
+			Subcomponents: []*types.Path{
+				&types.Path{
+					Origin: "openconfig",
+					Elem: []*types.PathElem{
+						{
+							Name: "components",
+						},
+						{
+							Name: "component",
+							Key: map[string]string{
+								"name":      "Ethernet1234",
+								"someField": "someValue",
+							},
+						},
+					},
+				},
+			},
+		}
+
+		_, err := sc.Reboot(ctx, req)
+		testErr(err, codes.InvalidArgument, "Xcvr is malformed", t)
```

Comment:
Is Xcvr a typo?

Reply:
Acknowledged, Xcvr is meant to stand in for transceivers, will update this